### PR TITLE
[SPARK-45954][SQL] Remove redundant shuffles

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR adds a new rule to remove redundant shuffles from the Spark plan if it and its child is not reused.

### Why are the changes needed?

Reduce shuffle to improve query performance. The use case is user add a repartition to increase parallelism for broadcast join, but this is useless for shuffle join.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Just a question. I thought this might introduce many plan improvements, but I don't see any plan updates in this PR. May I ask if you know why there is no plan changes in our test suite, @wangyum ?

**Comment by wangyum** [comment]
It should only take effect in the following cases, usually the user wants to increase the parallelism for `BroadcastHashJoin`/`BroadcastNestedLoopJoin`:
```scala
df1.repartition(num).join(df2, )
df1.repartition(num).select(xxx).join(df2, "id")
```

**Comment by ulysses-you** [comment]
`df1.repartition` is a user-specified shuffle, we should better respect it, or at least we should add a new config. e.g., it's likely people want to add a shuffle to solve skewed data before aggregation or something else.

**Comment by wangyum** [comment]
This change does not affect adding repartition before aggregation:
```scala
spark.table("t1").repartition(10).groupBy("a").count().explain(true)
```

```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- HashAggregate(keys=[a#16L], functions=[count(1)], output=[a#16L, count#23L])
   +- Exchange hashpartitioning(a#16L, 5), ENSURE_REQUIREMENTS, [plan_id=62]
      +- HashAggregate(keys=[a#16L], functions=[partial_count(1)], output=[a#16L, count#28L])
         +- Exchange RoundRobinPartitioning(10), REPARTITION_BY_NUM, [plan_id=58]
            +- FileScan parquet spark_catalog.default.t1[a#16L]
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 1]
```diff
@@ -68,7 +68,13 @@ case class EnsureRequirements(
...
         val numPartitions = distribution.requiredNumPartitions
           .getOrElse(conf.numShufflePartitions)
-        ShuffleExchangeExec(distribution.createPartitioning(numPartitions), child, shuffleOrigin)
+        // Avoid generating redundant ShuffleExchangeExec node
+        val shuffleChild = child match {
+          case s: ShuffleExchangeExec => s.child
+          case other => other
+        }
+        ShuffleExchangeExec(
+          distribution.createPartitioning(numPartitions), shuffleChild, shuffleOrigin)
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:77]: @ulysses-you It only takes effect if there are no operators between two shuffles.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:77]: My point is that, it's a user-specified shuffle, we can not just remove it. Does it affect shuffle reuse ? e.g.,

```
with t (select /*+ repartition */ * from t)
select * from t join t2
union all
select * from t
```

Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala]: Skip reuse shuffle case.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantShufflesSuite.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  test("Remove redundant shuffle") {
+    withTempView("t1", "t2") {
+      spark.range(10).select($"id").createOrReplaceTempView("t1")
+      spark.range(20).select($"id").createOrReplaceTempView("t2")
+      Seq(-1, 1000000).foreach { threshold =>
+        withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> threshold.toString) {
+          val query = spark.table("t1").repartition(10).join(spark.table("t2"), "id")
+          query.collect()
+          val shuffleNum = getShuffleExchangeNum(query)
+          if (threshold > 0) {
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantShufflesSuite.scala:47]: The test condition looks not good.
How about `threshold == 1000000`?

**Pull request closed without merging**